### PR TITLE
19 find minimum value from a list of numerical inputdata

### PR DESCRIPTION
**Pull request by Allukmon**
---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/find_minimum_value.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    
+    takes a list of numbers and return a single number which is the lowest value from the list.
+    
+    Parameters:
+        values: list, the input values to process
+        
+    Returns -> number: the lowest value from the list of numbers.
+    
+    Raises:
+        AssertionError: if input is not a list and numerical
```
Comment (Yevheniia-Rudenko) [inline solutions/find_minimum_value.py:18]: AssertionError: if input is not a list or contains non-numeric values.

File: solutions/find_minimum_value.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        
+    Examples:
+        >>> find_minimum_value([1, 2, 3, 4])
+        1
+        >>> find_minimum_value([2.24, 4, 5.23, 1])
+        1
+        >>> find_minimum_value([1.2, 3.4, 2.7, 0.5])
+        0.5
+    """
+    assert isinstance(values, list)  #verifies that 'values' is a list
```
Comment (Yevheniia-Rudenko) [inline solutions/find_minimum_value.py:28]: assert all(isinstance(i, (int, float)) for i in values), "All elements must be numeric."

Comment (Yevheniia-Rudenko) [reply solutions/find_minimum_value.py:28]: The function asserts that the input is a list, but it doesn't verify that the elements inside the list are numeric (either integers or floats). It's good to add that check to handle cases where non-numeric elements are in the list.

File: solutions/find_minimum_value.py (hunk 1 above)
Comment (Yevheniia-Rudenko) [reply solutions/find_minimum_value.py:18]: The phrase "numerical" in the Raises section could be more specific, like "if the elements of the list are not numeric."

File: solutions/find_minimum_value.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        1
+        >>> find_minimum_value([2.24, 4, 5.23, 1])
+        1
+        >>> find_minimum_value([1.2, 3.4, 2.7, 0.5])
+        0.5
+    """
+    assert isinstance(values, list)  #verifies that 'values' is a list 
+    ordered_list_of_values = sorted(values) #returns the list in ascending order
+    min = ordered_list_of_values[0]  #gives the minimum value from the ordered list
+    return min
```
Comment (Yevheniia-Rudenko) [inline solutions/find_minimum_value.py:31]: As an option, you could optimize the function by using Python's built-in min() function, which does the same thing but is more efficient and cleaner.

**Review by Yevheniia-Rudenko — COMMENTED** [review]
The implementation of the find_minimum_value function is solid and works as expected. The function correctly returns the minimum value from a list of numbers, and the test cases cover a wide range of scenarios, including edge cases and defensive checks.

I’ve added some comments to the code as part of the review. Please take a look and feel free to use the suggestions if you find them helpful. They mostly focus on simplifying some parts to make them more concise and clear, without changing the meaning. Let me know if you have any questions or need further clarification! 

In general, the code is clear and effective, and the tests provide good coverage of the function's expected behavior.

**Comment by Yevheniia-Rudenko** [comment]
It seems like you're encountering a CI check error related to the ruff action. The logs show that two files would be reformatted, but the process still fails with an exit code 1. I ran into the same issue, and once I figure out the solution in my example, I'll make sure to share it with you. In the meantime, you might want to check if there are any configuration adjustments that could resolve this error.

**Comment by Allukmon** [comment]
Thanks a lot. I will also try to look for solution on the CI check also.

Thanks a lot for the feedback 😊

**Comment by Yevheniia-Rudenko** [comment]
@Allukmon, I finally resolved my issue with CI checks, and based on that experience, I recommend first installing Ruff:
**pip install ruff**
You can check if it’s installed with:
**ruff --version**
Now, you should be able to use Ruff to format your files:
To format all files, use:
**ruff format**

File: solutions/find_minimum_value.py (hunk 3 above)
Comment (Allukmon) [reply solutions/find_minimum_value.py:31]: Thanks a lot for the suggestion I do not know that there is such a built-in function. Also, I wanted to see my problem-solving skills and my code-writing skills too. I would try that in my private repository.

Comment (Allukmon) [reply solutions/find_minimum_value.py:36]: I have updated the assertionError and statement. 

you might want to check it out

Comment (Yevheniia-Rudenko) [reply solutions/find_minimum_value.py:36]: Hi @Allukmon,

Thanks for sharing your thoughts! Using the min() function is just one of Python's handy features, but it's equally important to understand how things work under the hood, so I admire your approach!

 I checked the changes, and everything looks great!

**Review by Yevheniia-Rudenko — APPROVED** [review]
After reviewing the changes, I can confirm that the code looks great and aligns with the checklist! You've done a solid job implementing the updates. If you need any further feedback or assistance, feel free to reach out. Keep up the great work!

**Comment by Allukmon** [comment]
i was unable to install ruff on my machine but i was able to enable ruff using trunk extension.

you might want to help me if what i did does not reflect any changes🙏

**Comment by Yevheniia-Rudenko** [comment]
@Allukmon Thanks for your updates and the Trunk setup! I noticed several additional files in the push (e.g., .trunk/configs files, .trunk/trunk.yaml) that might not align with our project scope. To keep the repository clean:

Please review changes (git status / git diff) before pushing.
Remove unnecessary files using git rm.
Add irrelevant files to .gitignore to prevent future commits.
Additionally, instead of relying on Trunk, I recommend reinstalling ruff directly. I encountered an issue with ruff myself, and reinstalling it resolved the problem. This approach might simplify the setup and help avoid introducing extra configuration files.

**Comment by Allukmon** [comment]
@Yevheniia-Rudenko I have been able to pass all the CI-checks.

**Comment by Allukmon** [comment]
@Yevheniia-Rudenko, thanks for your help I been able to pass all the CI-checks

**Review by Yevheniia-Rudenko — APPROVED** [review]
Hi @Allukmon!

Thank you for your hard work and dedication. I know from experience that CI checks can be frustrating—great job handling them!

I am merging this branch into the main now.
